### ISSUE TITLE
new rule: require generic definitions when extending class/interface

### ISSUE THREAD

**Issue by martinheidegger** [typescript-eslint/typescript-eslint#2825]
TypeScript allows defaults for generics:

```typescript
class Foo <T extends any = any> {}
```

which is useful for instantiation but can become problematic when extending:

```typescript
class Bar extends Foo {}
class Baz extends Bar {} // How can Baz update T on Foo?
```

because in this case, the author of `Bar` may not be aware that it prevents any future narrowing of `T`. In my could I would like to be able to make sure that _if_ a type with a generic default value is extended, all generics must be set.

```typescript
class Bar extends Foo {} // Error: default T is not extended

class Bar2<T extends any = any> extends Foo<T> {} // valid; allowing future narrowing

class Bar3 extends Foo<any> {} // also valid; purposefully preventing future narrowing
```

It seems like a new rule `extend-no-generic-default` (naming?) could be a great addition to the eslint-plugin suite.

**Comment by bradzacher** [comment]
I don't quite understand - why do you *need* to forward-declare the generics?
When they're actually needed, you can just amend the declaration with it.

I don't think a lint rule really does anything here, apart from making you write unnecessary code.

**Comment by martinheidegger** [comment]
> you can just amend the declaration with it.

There is at least one common case where it is hard to amend a class: when the code is already published. As library writer I want to make sure that I don't forget to specify this in libraries.

**Comment by bradzacher** [comment]
I don't think that this rule belongs within this project - it is a very niche use-case for a general-purpose plugin like this.
You definitely could:
- create a new plugin for this and other similar rules specifically for maintaining libraries
- find a different, existing plugin that better fits this rule.
- implement the rule for use internal to your project.

**Comment by martinheidegger** [comment]
I believe that this would be beneficial to other users as well, and perhaps the ecosystem at large, but I can also see how this is effort that you don't want to undertake. Thank you for the consideration.
